Posts/2021/03/ScriptingGuy.0006.readingbottomup.md

```diff
@@ -0,0 +1,156 @@
+
+## Arrays vs text files
+
+There’s an important difference between a text file and an array.
+With a text file, using `Get-Content`, you read it from only from the start to the finish.
+Windows, .NET, and PowerShell do not provide a way to read the file in reverse.
+However, once you have the file contained in an array. it’s easy to read the array from the bottom to the top.
+
+Let's start by working out how many lines there are in the array.
+And, more as a sanity check, display how many lines there are in the file, like this:
+
+```powershell-console
+PS C:\Foo> $Array = Get-Content -Path C:\Foo\SmallFile.txt
+PS C:\Foo> $Length = $Array.count
+PS C:\Foo> "There are $Length lines in the file"
+There are 7 lines in the file
+```
+
+So that tells us you have the number of lines in the array that you expected. 
+
+## Getting Array Members
+
+So let's give you a solution. 
+In our sample array, `$Array` we have 7 lines.
+We can address directly any individual array member by using `[<index>]` syntax (after the array name).
+So the first item in the array always has an index number of 0 or `$Array[0`).
+In our array, the line **violet** has an index number of 0 so you can get to it using `$Array[0]`.
+Likewise, red has an index number of 6, or `$Array[6]`.
+But that doesn't help us much - just yet!
+
+A particularly neat feature of array handling in PowerShell is that we can work backwards in an array using negative index values.
+An index of [-1] is always the last element of an array, [-2] is the penultimate line, and so on.
+So `$Array[-1]` is Red, `$Array[-2]` is Orange, and so on.
+
+So what we do is to look first at `$Array[-1]`, then `$Array[-2]`, and so on, all withing a simple foreach loop, like this:
+
+```powershell-console
+PS C:\Foo> $Array = Get-Content -Path C:\Foo\SmallFile.txt
+PS C:\Foo> $Length = $Array.count
+PS C:\Foo> "There are $Length lines in the file"
+There are 7 lines in the file
+PS C:\Foo> $Line = 1
+PS C:\Foo> 1..$Length | foreach {$Array[-$Line]; $Line++}
+red
+orange
+yellow
+green
+blue
+indigo
+violet
+```
+
+This code snippet first sets a variable, `$Line`, to 1.
+Then you read the file and display how many lines are in the file.
+You then use `foreach` to run a script block once for each line in the file.
+Inside the script block you get the array element starting at the end and output it to the console.
+Then you increment the line number and repeat.
+
+This may be a little confusing if you haven't work with arrays, but once you get the hang of it, you'll see how simple it really is.
+Arrays are a fantastic capability within PowerShell.
+
+## For More Information
+
+For more information on arrays in PowerShell, see [About_Arrays](https://docs.microsoft.com/powershell/module/microsoft.powershell.core/about/about_arrays).
+And for more information on `Get-Content! see the [Get-Content Help page](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.management/get-content).
```

Comment:
```suggestion
And for more information on `Get-Content! see the [Get-Content]](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.management/get-content)  Help page.
```

Reply:
I made that change in the updated version.